cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -0,0 +1,604 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterapi
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/informers"
+	kubeinformers "k8s.io/client-go/informers"
+	kubeclient "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+	"k8s.io/utils/pointer"
+)
+
+const (
+	machineProviderIDIndex = "machineProviderIDIndex"
+	nodeProviderIDIndex    = "nodeProviderIDIndex"
+	defaultMachineAPI      = "v1alpha2.cluster.x-k8s.io"
```

Comment:
> Would it be worth using discovery to get the latest API version for the `cluster.x-k8s.io` group rather than using the constant here?

It should also be possible for an environment variable to override it at runtime/deployment time. For example, in OpenShift, we would want `v1beta1.machine.openshift.io`.

Reply:
I also mentioned a few other constants/annotations in https://github.com/gardener/autoscaler/issues/32#issuecomment-594406202

Reply:
We should be able to decouple the group variable from the preferred version, e.g:
```
        defaultAPIGroup := "cluster.x-k8s.io"
	groupList, err := discoveryClient.ServerGroups()
	if err != nil {
		return "", err
	}
	foundCAPIGroup := false
	var CAPIGroupVersion string
	for _, group := range groupList.Groups {
		if group.Name == "defaultAPIGroup" {
			foundCAPIGroup = true
			CAPIGroupVersion = group.PreferredVersion.Version
			break
		}
	}
	return CAPIGroupVersion
```

Similar to https://github.com/kubernetes/kubernetes/blob/fb6f5d739b08e049b188806d6e33099f85fad2fe/staging/src/k8s.io/kubectl/pkg/drain/drain.go#L96-L108

Probably reasonable to address in a follow up.

Reply:
just wanted to note that i am currently working on a downstream patch for using an environment variable to override the default version group. i can propose it here once we get things merged.

Reply:
we'll follow up with https://github.com/openshift/kubernetes-autoscaler/pull/134